File: core/framebuffer.nim
import image
import misc

type
  Framebuffer* = object
    width: int
    height: int
    color*: ptr uint8
    depth*: ptr float32

proc framebuffer_create*(w, h: int): Framebuffer =
  var color = create(uint8, w*h*4)
  var depth = create(float32, w*h)
  Framebuffer(width: w, height: h, color: color, depth: depth)

proc framebuffer_destroy*(fb: Framebuffer) =
  if fb.color != nil: dealloc(fb.color)
  if fb.depth != nil: dealloc(fb.depth)

proc width*(fb: Framebuffer): int = fb.width
proc height*(fb: Framebuffer): int = fb.height

proc framebuffer_clear*(fb: Framebuffer) =
    let size = fb.width * fb.height
    for i in 0..<size*4:
      fb.color[i] = 0

proc framebuffer_blit_rgb*(fb: Framebuffer, bitmap: Bitmap) =
    let size = fb.width * fb.height
    var buffer = bitmap.buffer
    for i in 0..<size*4:
        buffer[i] = fb.color[i]


proc framebuffer_blit_bgr*(fb: Framebuffer, bitmap: Bitmap) =
    let size = fb.width * fb.height
    var buffer = bitmap.buffer
    for i in 0..<size:
        for j in 0..<3:
            buffer[i*4+j] = fb.color[i*4+(2-j)]


proc framebuffer_blit_rgb*(fb: Framebuffer, bitmap: ptr Bitmap) =
    let size = fb.width * fb.height
    var buffer = bitmap.buffer
    for i in 0..<size*4:
        buffer[i] = fb.color[i]


proc framebuffer_blit_bgr*(fb: Framebuffer, bitmap: ptr Bitmap) =
    let size = fb.width * fb.height
    var buffer = bitmap.buffer
    for i in 0..<size:
        for j in 0..<3:
            buffer[i*4+j] = fb.color[i*4+(2-j)]

File: core/framebuffer.cpp
#include "framebuffer.h"

#include "image.h"
#include <stdlib.h>

framebuffer_t *framebuffer::create(int w, int h) {
  framebuffer_t *fb = (framebuffer_t *)(calloc(1, sizeof(framebuffer_t)));
  fb->width = w;
  fb->height = h;
  fb->color = (uint8_t *)calloc(w * h * 4, sizeof(uint8_t));
  fb->depth = (float *)calloc(w * h, sizeof(float));
  return fb;
}

void framebuffer::destroy(framebuffer_t *fb) {
  fb->width = 0;
  fb->height = 0;
  if (fb->color) {
    free(fb->color);
    fb->color = 0;
  }
  if (fb->depth) {
    free(fb->depth);
    fb->depth = 0;
  }
}

void framebuffer::clear(framebuffer_t *fb) {
  int size = fb->width * fb->height;
  for (int i = 0; i < size * 4; i++) {
    fb->color[i] = 0;
  }
}

void framebuffer::blit(framebuffer_t const *fb, bitmap_t *bitmap) {
#if defined(_WIN32)
  blit_bgr(fb, bitmap);
#elif defined(__APPLE__)
  blit_rgb(fb, bitmap);
#endif
}

void framebuffer::blit_rgb(framebuffer_t const *fb, bitmap_t *bitmap) {
  int size = fb->width * fb->height;
  for (int i = 0; i < size * 4; i++) {
    bitmap->buffer[i] = fb->color[i];
  }
}

// void blit_rgb(t_framebuffer fb, bitmap_t bitmap) {}
